backend/internal/th/ethereum.go

```diff
@@ -0,0 +1,224 @@
+package th
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"math/big"
+	"testing"
+
+	"github.com/Tangui-Bitfly/ethsimtracer"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/eth/ethconfig"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/ethereum/go-ethereum/rpc"
+
+	"github.com/gobitfly/beaconchain/internal/contracts"
+)
+
+type EOA struct {
+	*bind.TransactOpts
+	PrivateKey *ecdsa.PrivateKey
+}
+
+const simulatedChainID = 1337
+
+var (
+	OneEther            = big.NewInt(params.Ether)
+	DefaultTokenBalance = big.NewInt(100)
+)
+
+func CreateEOA(t *testing.T) *EOA {
+	t.Helper()
+
+	priv, err := crypto.GenerateKey()
+	if err != nil {
+		t.Fatal(err)
+	}
+	opts, err := bind.NewKeyedTransactorWithChainID(priv, big.NewInt(simulatedChainID))
+	if err != nil {
+		t.Fatal(err)
+	}
+	return &EOA{
+		PrivateKey:   priv,
+		TransactOpts: opts,
+	}
+}
+
+type BlockchainBackend struct {
+	*ethsimtracer.Backend
+	BankAccount *EOA
+	ChainID     int
+	Endpoint    string
+
+	rpc *rpc.Client
+}
+
+func NewBackend(t *testing.T, accounts ...common.Address) *BlockchainBackend {
+	t.Helper()
+
+	genesis := make(map[common.Address]types.Account)
+	for _, account := range accounts {
+		genesis[account] = types.Account{Balance: OneEther}
+	}
+	bankAccount := CreateEOA(t)
+	genesis[bankAccount.From] = types.Account{Balance: new(big.Int).Mul(big.NewInt(1000), OneEther)}
+
+	log.SetDefault(log.NewLogger(log.DiscardHandler()))
+	// TODO use random port
+	bk := ethsimtracer.NewBackend(genesis, func(nodeConf *node.Config, ethConf *ethconfig.Config) {
+		nodeConf.HTTPModules = append(nodeConf.HTTPModules, "debug", "eth")
+		nodeConf.HTTPHost = "127.0.0.1"
+		nodeConf.HTTPPort = 4242
```

Comment:
It might be a good idea to move the port value to the top of the file and store it as a constant - it will make it easier to update when needed